ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/ReexecutingExecutionPayloadBlockManager.java

```diff
@@ -0,0 +1,172 @@
+  private static final Logger LOG = LogManager.getLogger();
+  private static final int EXPIRES_IN_SLOT = 2;
+  private static final Duration RECHECK_INTERVAL = Duration.ofSeconds(2);
+
+  private final AtomicBoolean reexecutingExecutionPayload = new AtomicBoolean(false);
+  private final Set<SignedBeaconBlock> pendingBlocks = LimitedSet.create(10);
+  private final AsyncRunner asyncRunner;
+
+  private Optional<Cancellable> cancellable = Optional.empty();
+
+  private ReexecutingExecutionPayloadBlockManager(
+      RecentChainData recentChainData,
+      BlockImporter blockImporter,
+      PendingPool<SignedBeaconBlock> pendingBlocks,
+      FutureItems<SignedBeaconBlock> futureBlocks,
+      BlockValidator validator,
+      final AsyncRunner asyncRunner) {
+    super(recentChainData, blockImporter, pendingBlocks, futureBlocks, validator);
+    this.asyncRunner = asyncRunner;
+  }
+
+  public static BlockManager create(
+      final PendingPool<SignedBeaconBlock> pendingBlocks,
+      final FutureItems<SignedBeaconBlock> futureBlocks,
+      final RecentChainData recentChainData,
+      final BlockImporter blockImporter,
+      final BlockValidator validator,
+      final AsyncRunner asyncRunner) {
+    return new ReexecutingExecutionPayloadBlockManager(
+        recentChainData, blockImporter, pendingBlocks, futureBlocks, validator, asyncRunner);
+  }
+
+  @Override
+  public SafeFuture<?> doStart() {
+    return super.doStart()
+        .thenRun(
+            () ->
+                cancellable =
+                    Optional.of(
+                        asyncRunner.runWithFixedDelay(
+                            this::execute,
+                            RECHECK_INTERVAL,
+                            error -> LOG.error("Failed to reexecute execution payloads", error))));
+  }
+
+  @Override
+  protected SafeFuture<?> doStop() {
+    return super.doStop()
+        .thenRun(
+            () -> {
+              cancellable.ifPresent(Cancellable::cancel);
+              cancellable = Optional.empty();
+            });
+  }
+
+  @Override
+  public void onSlot(UInt64 slot) {
+    super.onSlot(slot);
+    pendingBlocks.removeIf(block -> block.getSlot().isLessThan(slot.plus(EXPIRES_IN_SLOT)));
+  }
+
+  @Override
+  public SafeFuture<BlockImportResult> importBlock(SignedBeaconBlock block) {
+    return super.importBlock(block)
+        .thenPeek(
+            blockImportResult -> {
+              if (requiresReexecution(blockImportResult)) {
+                pendingBlocks.add(block);
+              }
+            });
+  }
+
+  private void execute() {
+    if (reexecutingExecutionPayload.compareAndSet(false, true)) {
+      LOG.debug("re-executing execution payloads for queued blocks");
+      SafeFuture.allOfFailFast(
+              pendingBlocks.stream()
+                  .map(this::reexecuteExecutionPayload)
+                  .toArray(SafeFuture[]::new))
+          .thenPeek(__ -> reexecutingExecutionPayload.set(false))
```

Comment:
This won't get set to false if an error occurs.  I suspect we should move it to an `always` call at the end.